lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1007,6 +1006,10 @@ bool TypeChecker::coercePatternToType(Pattern *&P, TypeResolution resolution,
       // Assume the compiler knows what it's doing.
     } else if (diagTy->isEqual(Context.TheEmptyTupleType)) {
       shouldRequireType = true;
+    } else if (auto arrTy = constraints::ConstraintSystem::isArrayType(diagTy)) {
```

Comment:
I'd say let's just have `isArrayType` inlined here e.g.:

```c++
if (auto *BST = diagTy->getAs<BoundGenericStructType>()) {
  if (BST->getDecl() == Context.getArrayDecl())
    shouldRequireType = BST->getGenericArgs()[0]->isEqual(Context.TheEmptyTupleType);
}
```

As a follow-up it could be moved to `TypeChecker` along with `isDictionaryType` and `isSetType` as a static method.

Reply:
Sure, I'll make this change, thanks!